src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -513,11 +510,56 @@ private void UpdateStreamTimeouts(DateTimeOffset now)
         }
     }
 
+    private async Task CreateHttp3Stream<TContext>(ConnectionContext streamContext, Http3StreamContext context, IHttpApplication<TContext> application, long streamId) where TContext : notnull
+    {
+        // http request stream
+        // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-5.2-2
+        if (_gracefulCloseStarted)
+        {
+            // https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1.2-3
+            streamContext.Features.GetRequiredFeature<IProtocolErrorCodeFeature>().Error = (long)Http3ErrorCode.RequestRejected;
+            streamContext.Abort(new ConnectionAbortedException("HTTP/3 connection is closing and no longer accepts new requests."));
+            await streamContext.DisposeAsync();
+
+            return;
+        }
+
+        // Request stream IDs are tracked.
+        UpdateHighestOpenedRequestStreamId(streamId);
+
+        var persistentStateFeature = streamContext.Features.Get<IPersistentStateFeature>();
+        Debug.Assert(persistentStateFeature != null, $"Required {nameof(IPersistentStateFeature)} not on stream context.");
+
+        Http3Stream stream;
+
+        // Check whether there is an existing HTTP/3 stream on the transport stream.
+        // A stream will only be cached if the transport stream itself is reused.
+        if (!persistentStateFeature.State.TryGetValue(StreamPersistentStateKey, out var s))
+        {
+            stream = new Http3Stream<TContext>(application, context);
+            persistentStateFeature.State.Add(StreamPersistentStateKey, stream);
+        }
+        else
+        {
+            stream = (Http3Stream<TContext>)s!;
+            stream.InitializeWithExistingContext(streamContext.Transport);
+        }
+
+        _streamLifetimeHandler.OnStreamCreated(stream);
+        KestrelEventSource.Log.RequestQueuedStart(stream, AspNetCore.Http.HttpProtocol.Http3);
+        ThreadPool.UnsafeQueueUserWorkItem(stream, preferLocal: false);
+    }
+
     private async Task CreateAndAddWebTransportStream(Http3PendingStream stream, long streamId, WebTransportStreamType type)
     {
+        if (!_context.ServiceContext.ServerOptions.EnableWebTransportAndH3Datagrams)
```

Comment:
Change to `Debug.Assert(_context.ServiceContext.ServerOptions.EnableWebTransportAndH3Datagrams, "This should blksdfsdfsd")`